### PR TITLE
Introduce Constants should not be offered for Null Literals

### PR DESCRIPTION
**Pull request by balajikris**
Fixes internal bug 909152.

Don't offer Introduce Constants on NullLiterals. They could cause
semantic errors when used in a replace all occurrences context and
offer no value anyway.

This change is for C# only. VB works correctly.

### CODE REVIEW

**Comment by balajikris** [comment]
tagging for CR: @brettfo @basoundr @rchande @Pilchie

**Comment by rchande** [comment]
:+1:

File: src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs [hunk 1]
```diff
@@ -103,6 +103,11 @@ protected override bool CanIntroduceVariableFor(ExpressionSyntax expression)
                 return false;
             }
 
+            if (expression is LiteralExpressionSyntax && ((LiteralExpressionSyntax)expression).IsKind(SyntaxKind.NullLiteralExpression))
+            {
+                return false;
+            }
+
```
Comment (Pilchie) [inline src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs]: VB seems to have the same problem.

File: src/Features/Core/IntroduceVariable/AbstractIntroduceVariableService.State.cs [hunk 2]
```diff
@@ -216,9 +216,10 @@ private TExpressionSyntax GetExpressionUnderSpan(SyntaxTree tree, TextSpan textS
             private bool CanIntroduceVariable(
                 CancellationToken cancellationToken)
             {
-                // Don't generate a variable for an expression that's the only expression in a
-                // statement.  Otherwise we'll end up with something like "v;" which is not
-                // legal in C#.
+                // Don't generate a variable for an expression :
+                // (a) If that's the only expression in a statement.  
+                // Otherwise we'll end up with something like "v;" which is not legal in C#.
+                // (b) For Null Literals, as AllOccurences could introduce semantic errors.
```
Comment (Pilchie) [inline src/Features/Core/IntroduceVariable/AbstractIntroduceVariableService.State.cs]: I would move the comment about why checking for null is important to the place where you are testing for null.

File: src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs (hunk 1 above)
Comment (balajikris) [inline src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs]: doh, i trusted what the bug report said: *VB does not offer this refactoring for 'Nothing'* . Will check on VB.

Comment (balajikris) [inline src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs]: @Pilchie I could not reproduce this in VB. Do you have a repro? 
```VB
Class A
End Class

Class B
    Sub Test()
        Dim x As A = $Nothing
        Dim y As B = Nothing
    End Sub
End Class
```

Comment (Pilchie) [inline src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs]:
```VB
Module Program
    Sub Main(args As String())
        Main(Nothing)
        M(Nothing)
    End Sub

    Sub M(i As Integer)
    End Sub
End Module
```

Comment (balajikris) [inline src/Features/CSharp/IntroduceVariable/CSharpIntroduceVariableService.cs]: Yep, that does repro. Thanks.

**Comment by balajikris** [comment]
addressed VB in the latest commit. Looking for one more sign-off.

**Comment by Pilchie** [comment]
:+1:

**Comment by balajikris** [comment]
@basoundr, simplified that conditional, thanks!

**Pull request merged**